### DISCUSSION TITLE
Custom AsyncMethodBuilder does not work for custom types that take 2 or more parameters

### DISCUSSION THREAD

**Discussion by vsfeedback** [dotnet/csharplang#2273] [Language Ideas]
<p>Using a custom AsyncMethodBuilder with a typeof(ResultAsyncMethodBuilder&lt;,&gt;) over a class Result&lt;T1, T2&gt; gives CS1983 when used in a project:</p><p></p><p>public static async Result&lt;int, Exception&gt; Add(Result&lt;int, Exception&gt; first, Result&lt;int, Exception&gt; second) { </p><p>  var a = await first;</p><p>  ....</p><p>}</p><p></p><p>It would be nice to be be capable of using this for Result types, instead of just Option types</p>

_This issue has been moved from https://developercommunity.visualstudio.com/content/idea/465356/custom-asyncmethodbuilder-does-not-work-for-custom.html
VSTS ticketId: 797932_
_These are the original issue comments:_

Jane Wu [MSFT] on 2/22/2019, 02:52 AM (10 hours ago): <p>Thank you for taking the time to provide your suggestion. We will do some preliminary checks to make sure we can proceed further.&nbsp;We'll provide an update once the issue has been triaged by the product team.</p>

**Comment by jcouv** [comment]
Tagging @cston for triage

**Comment by YairHalberstadt** [comment]
Take a look at https://github.com/dotnet/csharplang/discussions/3403 for more on this topic
